File: Difficulty: Easy/Minimum Cost of ropes/minimum-cost-of-ropes.py
#User function Template for python3

from heapq import heapify,heappop,heappush
class Solution:
    def minCost(self,arr) :
        heapify(arr)
        ans = 0
        while len(arr)>=2:
            temp = heappop(arr)+heappop(arr)
            ans += temp
            heappush(arr,temp)
            
        return ans



#{ 
 # Driver Code Starts
#Initial Template for Python 3
import atexit
import io
import sys
import heapq
from collections import defaultdict
# Contributed by : Nagendra Jha

_INPUT_LINES = sys.stdin.read().splitlines()
input = iter(_INPUT_LINES).__next__
_OUTPUT_BUFFER = io.StringIO()
sys.stdout = _OUTPUT_BUFFER


@atexit.register
def write():
    sys.__stdout__.write(_OUTPUT_BUFFER.getvalue())


if __name__ == '__main__':
    test_cases = int(input())
    for cases in range(test_cases):
        a = list(map(int, input().strip().split()))
        ob = Solution()
        print(ob.minCost(a))

# } Driver Code Ends

File: Difficulty: Medium/Minimum Cost of ropes/minimum-cost-of-ropes.cpp
class Solution {
  public:
    int minCost(vector<int>& arr) {
        if(arr.size() == 1) return 0;
        int cost = 0, i = arr.size()-1;
        priority_queue<int, vector<int>, greater<int>> pq(arr.begin(), arr.end());
        
        while(i--){
            int a = pq.top(); pq.pop();
            int b = pq.top(); pq.pop();
            int temp = a+b;
            pq.push(temp);
            cost += temp;
        }
        return cost;
        
    }
};